### ISSUE TITLE
⚔Combat

### ISSUE THREAD

**Issue by peschkaj** [legitbiz/SealsOfFate#56]
- [x] Modify movement to detect collisions with player/enemies and initiate combat
- [x] Modify HP held in `GameManager`
- [x] Damage
  - [x] Player deals damage to enemies
  - [x] Enemies deal damage to the player

## Combat implementation

Create an attack object composed of effect components and modifier tags. Effects are things like "Do Damage" and "Knockback" while tags are "Element: Ice" or "Ignores Armor".

Damage can be applied to one or more stats (HP, mana, armor).

### How can we implement effects without a huge `if` statement?

- One option is to pass the attacked object into a method on the attack. This moves the `if` statement into the attack, but it still exists.
- Another option is to use a `switch` or `if` and limit ourselves to a minimal number of effects for now. It's also worth noting that these aren't inelegant.
- Treat effects and tags like a damage modification queue. The attacked object is passed into a `InflictDamage()` method that works through a copy of the attacked object's vital stats, modifies as appropriate based on the effect/tag, and either inflicts damage on the original object or else returns the amount of damage to be applied.

**Comment by Zonr0** [comment]
Effects won't be much of an issue as I originally envision this system working. Attacks will be built with a public list of effects that are iterated through on a successful attack. This should make it easy to apply a variety of effects, eg:

Some sample constructors might look like:
`EffectDamage(int hpDamage,int mpDamage=0)`
'EffectKnockback(int distance)'
'EffectForceMove(Vector2D moveVector) // Or a pair of x/y coordinates'
'EffectApplyStatus(StatusEffect toApply) // Eventually'
'EffectModifyStrengthTemporarily(int amount,int duration)'

Effect would be an ABC based loosely off the command pattern with
'void ApplyEffect(Character source, Character target)'
'void UndoEffect() //For temporary effects'

A bonus of this approach is that this system can be easily extended to consumable effects, traps, self-modifying spells, etc. It could possibly be the basis of a status effect system by extending them into a variant that are stored on the character and applied every turn.

Tags are the harder part to figure out. We could always start with ifs and switches at first, but I feel like if we want to add tags or interactions later, that's going to quickly become a pain. I like the idea of a damage modification queue, but that makes it difficult to apply tags like unblockable.

This of course is dependent on some decisions we make regarding combat, so this should be updated once we discuss on Sunday.

**Comment by Zonr0** [comment]
Also: another idea I had while talking to Jazzy: blubber should be effective against physical and ice damage, but makes you more vulnerable to fire. This forces a tradeoff where the more armor you have, the more resilient you are to most attacks, but the more vulnerable you are to more dangerous enemies. It could also, of course as connor suggested, make you slower.

**Comment by peschkaj** [comment]
> I like the idea of a damage modification queue, but that makes it difficult to apply tags like unblockable.

My grand idea is that combat involves a CombatData object. We transform the player or enemy into a CombatData that contains all information about what's happening right now - health, mana, armor, blocking.

Then walk through each effect and apply the effects, or do what's necessary to set up the knockback. So in the case of a seal blocking an unblockable ice attack that does 10 damage (in no particular order):

- Create a CombatData from the seal & enemy that contains the seal's health info, seal's defensive effects, enemy's attack effects, and enemy's tags.
- Execute the CombatData.ComputeDamage() to determine the actual damage.
- Attack's unblockable is applied, removing the seal's blocking modifier.
- Attack's damage is applied, but the application checks the defensive modifiers and finds that the Seal is 50% immune to cold, so damage is cut in half (10/2 = 5).
- CombatData.ComputeData() returns 5.

**Comment by connorkuehl** [comment]
I think a combination of the two trains of thought at play here could result in a highly extensible, easily implemented combat design. 

Playing off of Andrew's thoughts: we could implement _very_ general effects as classes in code. Classes like `Damage(int damage, int scale)`, `Displace(int x, int y)` (more general than knockback), `TemporarilyModifyStat(String statIdentifier, int mod, int duration)`

Once we have a fairly broad suite of effects, we can bundle them into a `CombatData` or `Attack` object like Jeremiah's suggested. That way one attack may damage the player but also slow them by temporarily modifying their movement stat, or something along those lines. 

The `CombatData` object can then `ComputeCombatEffect(EntityState es)`, invoking appropriate methods of the `EntityState` class (Could be an object that players and enemies both have which contain information like health, movement abilities, etc). Maybe not even methods, maybe just properties, whatever. It would just iterate through all of the `Effects` it has contained, applying the appropriate modifications to the `EntityState` variables.

Now, the fun part. Adding and removing combat capabilities to the game could be done in an easily parsed domain language. We would need some sort of parser to produce an `Attack` from each scripted line of our domain specific language. 

```
"Slam attack": damage, displace, temporarilymodifystat
"Melee": damage
"Bull rush": damage, displace
```

For example, an `Attack` object would be created for BullRush, and it has a damage component and a displacement component. When that `Attack` object is processed over an `EntityStats`, the Entity's health will be reduced from the damage and the Entity's position will be moved due to displacement. In fact, an `Attack` is just a collection of `Effects`.

Obviously this is a rough concept, and the domain specific language may need more information about the damage numbers, and scale, etc. 

What do you guys think?

**Comment by peschkaj** [comment]
:+1:

I ended up going down this route eventually on my branch. I created some general purpose things like an `AttackInfo` to hold the damage and type. 

I like the way you fleshed this out with Andrew's ideas - it _should_ be pretty easy to combine everything into a general purpose set of combat functionality that isn't difficult to implement.

**Comment by Zonr0** [comment]
I like these ideas! I forgot to mention that there should be an Attack object that is composed of effects so I'm glad you guys picked that up and extended on it.

I like the idea of the domain language, but I think we'll also be able to get a lot of mileage of just attaching script references in the Unity editor.

OTOH, in the future well past mvp, we could use the scripting to provide a list of tags and situations in which an enemy should use an attack which could be use to implement a simple fuzzy logic attack selector in the AI.

**Comment by connorkuehl** [comment]
Of course! I'm glad you mentioned adding them as components to an Attack prefab. I'm still not well-practiced with the Unity editor and sometimes forget those details can be handled in the editor itself. That'd probably be the most Unity-idiomatic way to do it and if for whatever reason we come up on a wall we can switch easily, I think.

**Comment by peschkaj** [comment]
A basic combat resolution system was implemented in #69

**Comment by peschkaj** [comment]
We're currently blocked on getting custom UI for the following:
- [ ] CombatData
- [ ] AttackInfo
- [ ] AttackTag
- [ ] Damage
- [ ] DefenseInfo
- [ ] DefenseTag
